zcash_primitives/src/note_encryption.rs

```diff
@@ -250,18 +257,41 @@ pub struct SaplingNoteEncryption {
     to: PaymentAddress<Bls12>,
     memo: Memo,
     ovk: OutgoingViewingKey,
+    rseed: Option<[u8; 32]>,
 }
 
 impl SaplingNoteEncryption {
-    /// Creates a new encryption context for the given note.
-    pub fn new<R: RngCore + CryptoRng>(
+    /// Creates a new encryption context for the given note (before Canopy).
+    pub fn new_v1<R: RngCore + CryptoRng>(
```

Comment:
we want `new_v1()` to be deterministic. define `generate_v1(rng: &mut R)` which calls `new_v1()`. similarly for `generate_v2()`, `new_v2()`.